### PR TITLE
net: connection: Allow same port number for dest and source

### PR DESCRIPTION
**Pull request by jukkar**
Fix regression and allow incoming packet when source and
destination port numbers are the same.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

### CODE REVIEW

File: subsys/net/ip/connection.c [hunk 1]
```diff
@@ -794,7 +778,22 @@ static bool is_invalid_packet(struct net_pkt *pkt,
...
+	} else if (IS_ENABLED(CONFIG_NET_IPV6) &&
+		   net_pkt_family(pkt) == AF_INET6) {
+		if (net_ipv6_addr_cmp(&ip_hdr->ipv6->src,
+				      &ip_hdr->ipv6->dst) ||
+		    net_ipv6_is_my_addr(&ip_hdr->ipv6->src)) {
+			my_src_addr = true;
+		}
+	}
+
+	return my_src_addr && (src_port == dst_port);
```
Comment (tbursztyka) [inline subsys/net/ip/connection.c]: That won't work: if src_port != dst_port, this function should return true. The check on the ip address sources is only valid if and only if src_prt == dst_port.

Comment (dgloeck) [reply subsys/net/ip/connection.c]: No, if src_port != dst_port, this function should return false.

I guess the point of moving the src_port != dst_port check to the top in 1be2706a93f1ae6 was to speed up the most common case.

Comment (jukkar) [reply subsys/net/ip/connection.c]: This function should check that if
* we are receiving a localhost network packet and
* that the port numbers are the same one
then that is an invalid packet. Checking just src != port at the beginning is not enough.

Comment (tbursztyka) [reply subsys/net/ip/connection.c]: was reading it the other way round, nevermind

**Comment by jukkar** [comment]
Fixed indentation

File: subsys/net/ip/connection.c (hunk 1 above)
Comment (dgloeck) [reply subsys/net/ip/connection.c]:
```
return X && Y
```
is equal to
```
if (!Y)
  return false;
return X;
```

**Review by marc-hb — COMMENTED** [review]
If changing half the code then might as well go one step further and rename it to `is_valid_packet()` (or anything not negative) and remove all the double negations littered all over.

**Comment by jukkar** [comment]
> If changing half the code then might as well go one step further and rename it to `is_valid_packet()` (or anything not negative) and remove all the double negations littered all over.

Calling the function "invalid" or "valid" is pretty much the same thing, just removing or adding "!", but sure that can be done. I am more interested in if this change looks ok to everyone.

**Comment by pfalcon** [comment]
> Calling the function "invalid" or "valid" is pretty much the same thing

I'd be with @marc-hb here, if it leads to such confusion as discussion in a couple of PRs show, I'd suggest to simplify it for us human beings by removing one level of nestedness.

**Comment by jukkar** [comment]
Renamed `is_invalid_packet()` to `is_valid_packet()`

**Review by pfalcon — APPROVED** [review]

**Review by rveerama1 — APPROVED** [review]

**Comment by marc-hb** [comment]
Renaming `my_src_addr` to something like `same_addr` or `local_packet` would be more accurate considering you test not just "my source address". It would turn the final return statement into something that would be a bit English-looking at last hence not needing comments:
```
  return not (same_addr and same_port)
````

**Comment by pfalcon** [comment]
@galak: Can we please merge this, as I believe this is on critical path for performing testing and other fixes.

**Pull request merged**